File: csharp/Lib/72.EditDistance.cs
/*
72. Edit Distance
*/

using System.Linq;

public class EditDistance {
    public int MinDistance(string word1, string word2)
    {
		if (string.IsNullOrEmpty(word1) && !string.IsNullOrEmpty(word2))
			return word2.Length;
		if (!string.IsNullOrEmpty(word1) && string.IsNullOrEmpty(word2))
			return word1.Length;
		if (string.IsNullOrEmpty(word1) && string.IsNullOrEmpty(word2))
			return 0;
		
		return _MinDistance(word1, word2);
    }

	private int[][] _GenTable(int numRows, int numCols) {
		var table = new int[numRows][];

		for (int i = 0; i < table.Length; ++i) {
			table[i] = new int[numCols];
		}

		table[0][0] = 0;

		return table;
	}

	private int _MinDistance(string str1, string str2) {
		var dp = _GenTable(str1.Length + 1, str2.Length + 1);

		for (int i = 0; i < dp.Length; ++i)
			for (int j = 0; j < dp[i].Length; ++j)
				if (i == 0) dp[i][j] = j;
				else if (j == 0) dp[i][j] = i;
				else if (str1[i-1] == str2[j-1]) dp[i][j] = dp[i-1][j-1];
				else dp[i][j] = 1 +
					new int[] {
						dp[i-1][j],
						dp[i-1][j-1],
						dp[i][j-1],
					}.Min();
	
		return dp.Last().Last();
	}
}

File: python/72.EditDistance.py
# 72. Edit Distance

from dataclasses import dataclass

class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        if not word1 and not word2:
            return 0

        if not word1:
            return len(word2)
        if not word2:
            return len(word1)

        dp = []

        for i in range(0, len(word1) + 1):
            dp.append([])

            for j in range(0, len(word2) + 1):
                val = 0
                if i == 0: val = j
                elif j == 0: val = i
                elif word1[i-1] == word2[j-1]: val = dp[i-1][j-1]
                else: val = 1 + min(dp[i-1][j], dp[i-1][j-1], dp[i][j-1])

                dp[i].append(val)

        return dp[-1][-1]

@dataclass
class MetaData:
    word1: str
    word2: str


if __name__ == "__main__":

    sol = Solution()

    inputs = [
        MetaData("horse", "ros"),
        MetaData("intention", "execution")
    ]

    for input in inputs:
        print(input.__str__() + ": " + sol.minDistance(input.word1, input.word2).__str__())
        